Add reviews

SquareSweet opened the pull request.

SquareSweet committed:
rebase onto develop

SquareSweet committed:
<feat> - add reviews

SquareSweet committed:
<feat> - add reviews

SquareSweet committed:
<feat> - add review rates

SquareSweet committed:
<feat> - add review rates

SquareSweet committed:
rebase onto develop

SquareSweet committed:
<fix> - review rate display when getting all reviews

SquareSweet committed:
<fix> - review rate display when getting all reviews

SquareSweet committed:
<fix> - review rate display when getting all reviews

SquareSweet committed:
<test> - add reviews tables to all test schemas

SquareSweet committed:
<fix> - add @Transactional annotations for likes & dislikes interaction methods

SquareSweet committed:
<fix> - typo in sql request

SquareSweet committed:
<test> - repositories tests

SquareSweet committed:
<fix> - fixed throwing wrong exceptions on likes & dislikes interactions

SquareSweet committed:
<test> - review service tests

SquareSweet committed:
Merge remote-tracking branch 'origin/add-reviews' into add-reviews

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/repository/impl/ReviewRateRepositoryJdbcImpl.java
#	src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java
#	src/main/resources/schema.sql

pol-yap commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/ReviewRateRepositoryJdbcImpl.java:
```diff
@@ -0,0 +1,52 @@
+        return ReviewRate.builder()
+                .user(user)
+                .review(review)
+                .build();
```
isPositive нет в возвращаемом ReviewRate. Вероятно, он там никому особо и не нужен, но как-то неаккуратненько)

pol-yap commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,123 @@
+                        .map(ReviewRate::getUser)
+                        .collect(Collectors.toList())))
+                .collect(Collectors.toList());
+    }
```
N+1 deteceted)) Может, стянуть сразу все reviewRates в один лист и в стриме рассовать по review?

Difirton commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,82 @@
+    @GetMapping
+    public List<Review> getReviewByFilmId(@RequestParam(name = "count") Optional<Integer> count,
+                                          @RequestParam(name = "filmId") Optional<Long> filmId) {
+        if (filmId.isPresent()) {
```
Предлагаю в сервисе создать отдельный метод и отдать оба Optional в сервис и там уже сделать их распаковку и решить какой из методов сервиса будет вызван.

pol-yap replied:
В ветке add-most-popular тоже появляется похожая ситуация. И меня смущает, что IDEA подсвечивает Optional в параметрах. Типа, не рекомендуется так делать. Надо сформировать отношение к этому)) И либо забить и спокойно передавать Optional из метода в метод, либо сразу их распаковывать.

Difirton replied:
Я не в идее смотрел, но я не разу не встречал историю, что Optional отдавать в метод это плохо. Вот null отдавать вот это плохо, даже очень плохо, а Optional, наоборот норм. Просто у нас контроллеры в проекте - худые и чисты от всякой логики, просто занимаются обработкой запросов и вызовом сервисов.

Difirton commented on src/main/java/ru/yandex/practicum/filmorate/error/ReviewNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ReviewNotFoundException extends RuntimeException {
```
В GlobalExceptioHandler (пакет error) нужно зарегистрировать данный эксепшен. Во время рефакторинга разделю на два пакета, раз не увидела хендлер, значит это уже мой косяк, что он никак не выделен=)

Difirton commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/ReviewRateRepositoryJdbcImpl.java:
```diff
@@ -0,0 +1,52 @@
+    private static final String SQL_SELECT_BY_REVIEW_ID = "SELECT * FROM users_rates_reviews WHERE review_id = ?";
+    private static final String SQL_UPDATE_REVIEW_USEFUL = "UPDATE reviews SET useful = useful + ? WHERE id = ?";
+
+
```
Лишняя строчка=)

Difirton commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/ReviewRateRepositoryJdbcImpl.java:
```diff
@@ -0,0 +1,52 @@
+    public ReviewRate save(User user, Review review, boolean isPositive) {
+        this.jdbcOperations.update(SQL_INSERT_ALL_FIELDS, user.getId(), review.getId(), isPositive);
+        this.jdbcOperations.update(SQL_UPDATE_REVIEW_USEFUL,
+                (isPositive ? 1  : -1),
```
Я не очень люблю тернарные операторы, но тут сделано красиво=)

Difirton commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,123 @@
+
+    public List<Review> getAllReviews(Integer count) {
+        return reviewRepository.findAll(count).stream()
+                .peek(r -> r.setUsersRates(reviewRateRepository.getByReviewId(r.getId())
```
.peek() - метод существует для отладки, вот выкопировка из java doc:
API Note:
This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline:
Предлагаю подобрать другие варианты, даже если потребуются дополнительные расходы памяти, это лучше чем код с сайд-эффектами.

pol-yap replied:
хм.. тоже не раз слышал, что использовать peek не очень здорово, но всякий раз без упоминания почему. А что за сайд-эффекты?

Difirton replied:
Стримы это одна из реализаций функционального программировпния в Java. Функциональное программирование накладывает ограничения на сохранение состояния объекта. На эту тему есть очень крутая книга: "Современный язык Java" за авторством Урма, Фуско и Майкрофта. Лучше стрим использовать по его прямому назначению, как конвейер. Сайд эффекты будут возникать, когда появиться многопоточка, это когда возникает гонка за данными и не возможно реально предсказать состояние данных. Peek провоцирует это состояние, хоть для нас это не актуально, но лучше его использовать по прямому назначению: только для отладки, как рекомендует Java Doc. Ещё нам на хард вебинаре по стрима об этом говорили=)

SquareSweet replied:
Как вариант, сделать циклом, как при подгрузке жанров и режиссёров в фильмы?
И без .peek(), и похожие задачи будут сделаны по одному принципу.

pol-yap replied:
Кстати да, не сошёлся же свет клином на стримах)

Difirton replied:
Тоже против цикла for each не вижу ничего плохого. Peek просто ревьюер не примет, если увидит)

Difirton reviewed (commented):
Хорошая работа! Оставил пару комментариев.

SquareSweet committed:
<fix> - solved N+1 problem when getting review rates

pol-yap commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -45,17 +45,21 @@ public Review createReview(Review review) {
+                        .filter(rate -> rate.getReview().getId().equals(review.getId()))
                         .map(ReviewRate::getUser)
                         .collect(Collectors.toList())))
                 .collect(Collectors.toList());
```
Тоже тогда, может, на цикле сделать, чтоб уйти от этого грешного peek()?

SquareSweet replied:
Да, уже в процессе. Я сделала коммит до того, как увидела новые комментарии)

SquareSweet committed:
<fix> - move Optional handling from review controller to review service
<fix> - add ReviewNotFoundException to GlobalExceptionHandler
<refactor> - avoid peek() usage when getting review rates

pol-yap reviewed (approved):
По-моему, всё норм)

Difirton reviewed (approved):
Отличная работа.=)
